### PR TITLE
Update verifyoracledataresult in bridge to contain more information

### LINKED ISSUE #626: Update VerifyOracleDataResult in bridge to contain more information
See #625 for information that gets added.

### PR DESCRIPTION
**Pull request by prin-r**
Fix #626

### CODE REVIEW

File: bridges/evm/contracts/BandChainLib.sol [hunk 1]
```diff
@@ -1,6 +1,16 @@
 pragma solidity 0.5.14;
 
 library BandChainLib {
+
+    struct Result {
```
Comment (taobun) [inline bridges/evm/contracts/BandChainLib.sol]: Can remove this struct?

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/BandChainLib.sol (hunk 1 above)
Comment (prin-r) [reply bridges/evm/contracts/BandChainLib.sol]: done

File: bridges/evm/contracts/IBridge.sol [hunk 2]
```diff
@@ -2,9 +2,19 @@ pragma solidity 0.5.14;
...
+    uint64 requestedValidatorsCount;
+    uint64 sufficientValidatorCount;
+    uint64 reportedValidatorsCount;
+    bytes data;
+  }
+
   /// Helper struct to help the function caller to decode oracle data.
   struct VerifyOracleDataResult {
-    bytes data;
+    WrappedResult result;
```
Comment (sorawit) [inline bridges/evm/contracts/IBridge.sol]: I personally think this is confusing. Why don't we have `VerifyOracleDataResult` containing all the fields? like this.

```
struct VerifyOracleDataResult {
  uint64 oracleScriptId;
  uint64 requestTime;
  uint64 aggregationTime;
  uint64 requestedValidatorsCount;
  uint64 sufficientValidatorCount;
  uint64 reportedValidatorsCount;
  bytes params;
  bytes data;
}
```

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/IBridge.sol (hunk 2 above)
Comment (prin-r) [reply bridges/evm/contracts/IBridge.sol]: done

File: bridges/evm/contracts/Bridge.sol [hunk 3]
```diff
@@ -99,6 +99,32 @@ contract Bridge is IBridge {
...
+    bytes memory data = new bytes(_data.length - 40);
+    uint256 l = ((data.length - 1) / 32) + 1;
+    for (uint256 i = 0; i < l; i++) {
+      assembly {
+        mstore(add(data,add(0x20, mul(i,0x20))), mload(add(_data, add(0x48, mul(i,0x20)))))
+      }
+    }
+    result.data = data;
+
+    return  result;
```
Comment (taobun) [inline bridges/evm/contracts/Bridge.sol]: space?

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/Bridge.sol (hunk 3 above)
Comment (prin-r) [reply bridges/evm/contracts/Bridge.sol]: Right now we are using auto solidity format

File: bridges/evm/contracts/Bridge.sol [hunk 4]
```diff
@@ -113,6 +113,54 @@ contract Bridge is IBridge {
         bytes32 dataHash;
     }
 
+    /// Decode the encoded result and receives back the decoded data which is the data and its context.
```
Comment (sorawit) [inline bridges/evm/contracts/Bridge.sol]: Decodes ... returns back

File: bridges/evm/contracts/Bridge.sol [hunk 5]
```diff
@@ -113,6 +113,54 @@ contract Bridge is IBridge {
...
+        pure
+        returns (VerifyOracleDataResult memory)
+    {
+        require(_data.length > 40, "INPUT_MUST_BE_LONGER_THAN_40_BYTES");
+
+        VerifyOracleDataResult memory result;
+        assembly {
+            mstore(
+                add(result, 0x20),
+                and(mload(add(_data, add(0x08, 0x00))), 0xffffffffffffffff)
```
Comment (sorawit) [inline bridges/evm/contracts/Bridge.sol]: Having `add(0x08, 0x00)` is weird. Why don't we just use `0x08`? Ditto below.

File: bridges/evm/contracts/Bridge.sol [hunk 6]
```diff
@@ -113,6 +113,54 @@ contract Bridge is IBridge {
         bytes32 dataHash;
     }
 
+    /// Decode the encoded result and receives back the decoded data which is the data and its context.
+    /// @param _data The encoded of result and its context.
+    function decodeResult(bytes memory _data)
```
Comment (sorawit) [inline bridges/evm/contracts/Bridge.sol]: Let's call the parameter `_encodedData`

File: bridges/evm/contracts/Bridge.sol [hunk 7]
```diff
@@ -113,6 +113,54 @@ contract Bridge is IBridge {
...
+                and(mload(add(_data, add(0x08, 0x18))), 0xffffffffffffffff)
+            )
+            mstore(
+                add(result, 0xa0),
+                and(mload(add(_data, add(0x08, 0x20))), 0xffffffffffffffff)
+            )
+        }
+
+        bytes memory data = new bytes(_data.length - 40);
+        uint256 l = ((data.length - 1) / 32) + 1;
```
Comment (sorawit) [inline bridges/evm/contracts/Bridge.sol]: Call this `dataLengthInWords`

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/Bridge.sol [hunk 8]
```diff
@@ -113,6 +113,54 @@ contract Bridge is IBridge {
         bytes32 dataHash;
     }
 
+    /// Decode the encoded result and returns back the decoded data which is the data and its context.
```
Comment (taobun) [inline bridges/evm/contracts/Bridge.sol]: Decodes?

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/Bridge.sol (hunk 8 above)
Comment (prin-r) [reply bridges/evm/contracts/Bridge.sol]: done

File: bridges/evm/contracts/Bridge.sol [hunk 9]
```diff
@@ -162,11 +210,16 @@ contract Bridge is IBridge {
...
+        VerifyOracleDataResult memory result = decodeResult(_data);
+        result.params = _params;
+        result.oracleScriptId = _oracleScriptId;
+
+        return result;
     }
 
     /// Performs oracle state relay and oracle data verification in one go. The caller submits
-    /// the encoded proof and receives back the decoded data, ready to be validated and used.
+    /// the encoded proof and returns back the decoded data, ready to be validated and used.
```
Comment (sorawit) [inline bridges/evm/contracts/Bridge.sol]: Why changed this?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/Bridge.sol (hunk 9 above)
Comment (prin-r) [reply bridges/evm/contracts/Bridge.sol]: done

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**